server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -189,24 +191,57 @@ private void executeKnnVectorQuery(SearchContext context) throws IOException {
         List<DfsKnnResults> knnResults = new ArrayList<>(knnVectorQueryBuilders.size());
         for (int i = 0; i < knnSearch.size(); i++) {
             Query knnQuery = searchExecutionContext.toQuery(knnVectorQueryBuilders.get(i)).query();
-            TopScoreDocCollector topScoreDocCollector = TopScoreDocCollector.create(knnSearch.get(i).k(), Integer.MAX_VALUE);
-            CollectorManager<Collector, Void> collectorManager = new SingleThreadCollectorManager(topScoreDocCollector);
-            if (context.getProfilers() != null) {
-                InternalProfileCollectorManager ipcm = new InternalProfileCollectorManager(
-                    new InternalProfileCollector(collectorManager.newCollector(), CollectorResult.REASON_SEARCH_TOP_HITS)
-                );
-                QueryProfiler knnProfiler = context.getProfilers().getDfsProfiler().addQueryProfiler(ipcm);
-                collectorManager = ipcm;
-                // Set the current searcher profiler to gather query profiling information for gathering top K docs
-                context.searcher().setProfiler(knnProfiler);
-            }
-            context.searcher().search(knnQuery, collectorManager);
-            knnResults.add(new DfsKnnResults(topScoreDocCollector.topDocs().scoreDocs));
+            knnResults.add(singleKnnSearch(knnQuery, knnSearch.get(i).k(), context.getProfilers(), context.searcher()));
+        }
+        context.dfsResult().knnResults(knnResults);
+    }
+
+    /**
+     * @param <C> the type of collector
+     * @param <R> the #reduce() return type
+     */
+    static class CollectorManagerAdapter<C extends Collector, R> implements CollectorManager<C, Void> {
+
+        private final CollectorManager<C, R> wrapped;
+        private final Consumer<R> resultConsumer;
+
+        CollectorManagerAdapter(CollectorManager<C, R> in, Consumer<R> resultConsumer) {
+            this.wrapped = in;
+            this.resultConsumer = resultConsumer;
         }
+
+        @Override
+        public C newCollector() throws IOException {
+            return wrapped.newCollector();
+        }
+
+        @Override
+        public Void reduce(Collection<C> collectors) throws IOException {
+            resultConsumer.accept(wrapped.reduce(collectors));
+            return null;
+        }
+    }
+
+    static DfsKnnResults singleKnnSearch(Query knnQuery, int k, Profilers profilers, ContextIndexSearcher searcher) throws IOException {
+        final TopDocs[] topDocs = new TopDocs[1];
+        CollectorManager<? extends Collector, Void> cm = new CollectorManagerAdapter<>(
+            TopScoreDocCollector.createSharedManager(k, null, Integer.MAX_VALUE),
+            res -> topDocs[0] = res
+        );
+
+        if (profilers != null) {
+            ProfileCollectorManager ipcm = new ProfileCollectorManager(cm, CollectorResult.REASON_SEARCH_TOP_HITS);
```

Comment:
Would you like to play, as a follow-up, to make the profile collector manager typed, so that it can return the same result as the collector manager it wraps? That way its reduce could reduce the profile outputs, but also return the top docs, and you would not need that adapter? I think that would help me as well later. have you tried this already and maybe hit blockers?